feat: add firework shows

Sqrl34 opened the pull request:
## Description

Fireworks show for the 4th. Add multiple locations with multiple scenes with various events all at the same time. Shows started through a command and a permission system.

Known bug: The object does not spawn when doing shows but does spawn with individual fireworks.

## Checklist

<!-- Put an x inside the [ ] to check an item, like so: [x] -->

- [x] I have personally loaded this code into an updated Qbox project and checked all of its functionality.
- [x] My pull request fits the contribution guidelines & code conventions.

Sqrl34 committed:
feat: add firework shows

fireworks for the 4th!

Sqrl34 committed:
fixed scenes

Sqrl34 committed:
added permission system

Sqrl34 committed:
fixed space

Manason commented on config/shared.lua:
```diff
@@ -49,4 +49,115 @@ return {
+
+    shows = {
+        ['lsia'] = { -- location for command
+            { -- Sequences
```
What's the benefit of having difference sequences versus one show being one sequence?

Sqrl34 replied:
The benefit of having different sequences just for visual separation between different segments in a show. If you really wanted to you could just have 1 sequence with a bunch of events with a longer wait time but with sequences, I feel it is easier to divide the show into different blocks.

Manason replied:
Couldn't the same thing be accomplished with comments? I would get rid of sequences as a data type in favor of a more simple representation of the data

Sqrl34 replied:
sequences have been removed, to achieve same effect just increase a wait time

Manason commented on client/main.lua:
```diff
@@ -75,3 +105,17 @@ lib.callback.register('qbx_fireworks:client:useFirework', function(asset)
 end)
+
+RegisterNetEvent('qbx_fireworks:client:startShow', function(show)
+    if show ~= nil then
+        for i=1, #sharedConfig.shows[show] do -- Sequences
+            for j=1, #sharedConfig.shows[show][i].fireworks do -- Individual
+                TriggerServerEvent('qbx_fireworks:server:spawnShowObject', sharedConfig.shows[show][i].fireworks[j].asset, sharedConfig.shows[show][i].fireworks[j].coords)
+                startFireworkShow(sharedConfig.shows[show][i].fireworks[j].asset, sharedConfig.shows[show][i].fireworks[j].coords, sharedConfig.shows[show][i].fireworks[j].height)
+                Wait(sharedConfig.shows[show][i].fireworks[j].wait)
+            end
+            Wait(sharedConfig.shows[show][i].masterWait)
+        end
+    end
```
using variables will make the code more readable. Avoid referencing the same table value twice.
https://manason.github.io/effective-fivem-lua/tables/#extract-duplicate-table-dereferences-into-local-variables

Sqrl34 replied:
outdated

Manason commented on client/main.lua:
```diff
@@ -75,3 +105,17 @@ lib.callback.register('qbx_fireworks:client:useFirework', function(asset)
 end)
+
+RegisterNetEvent('qbx_fireworks:client:startShow', function(show)
+    if show ~= nil then
```
https://manason.github.io/effective-fivem-lua/functions/#use-guard-clauses

Sqrl34 replied:
outdated

Sqrl34 committed:
made it more readable

Sqrl34 commented:
added 2 new locales to the en.json so that will have to be updated in the other languages to

Manason reviewed (changes requested):
Maybe the show should be server based rather than client based? What happens if the starting client disconnects in the middle of it? What is responsible for showing fireworks on other clients? Right now it looks as if only the starting client would see fireworks

Sqrl34 committed:
revised initial request

Made the shows server sided. Removed sequences from the config.

Tests Ran:
1. Player 1 in paleto and Player 2 at lsia -> player 2 sees/hears fireworks and player 1 hears faint explosion

2. Player 1 starts fireworks show and Player 2 remains in game -> show continues to go on regardless of show initiator leaving

3. Player 1 starts show and Player 2 joins server while show is in progress -> Player 2 hears/sees current show happening

4. Player 1 starts show and teleports away while player 2 remains -> show continues as normal

Sqrl34 commented:
Tests Ran from the above revision
1. Player 1 in paleto and Player 2 at lsia -> player 2 sees/hears fireworks and player 1 hears faint explosion

2. Player 1 starts fireworks show and Player 2 remains in game -> show continues to go on regardless of show initiator leaving

3. Player 1 starts show and Player 2 joins server while show is in progress -> Player 2 hears/sees current show happening

4. Player 1 starts show and teleports away while player 2 remains -> show continues as normal

Manason commented on server/main.lua:
```diff
@@ -44,9 +44,20 @@ end)
 lib.addCommand('startshow', {
```
Should we expose this function as an export as well?

Sqrl34 replied:
possibly, added export

Manason commented on server/main.lua:
```diff
@@ -44,9 +44,20 @@ end)
-    TriggerClientEvent('qbx_fireworks:client:startShow', source, args.Location)
+    local show = args.Location
+    if not show then return end
+    if not config.shows[show] then return end
+    for i = 1, #config.shows[show].fireworks do -- Sequences
```
Consider extracting config.shows[show] to a variable.

Sqrl34 replied:
outdated

Manason commented on server/main.lua:
```diff
@@ -44,9 +44,20 @@ end)
+    if not config.shows[show] then return end
+    for i = 1, #config.shows[show].fireworks do -- Sequences
+        local firework = config.shows[show].fireworks[i]
+        local fireworkAsset = firework.asset
+        local fireworkCoords = firework.coords
+        local fireworkHeight = firework.height
+        local fireworkWait = firework.wait
```
Not sure there is any readability gain by storing these variables

Sqrl34 replied:
removed variables

Manason reviewed (changes requested):
Could more of the client show logic be moved to the server? This is still setup so than a single client still owns the show.

What happens if a client connects in the middle of a show?

Sqrl34 committed:
added export and made fireworks server sided

Sqrl34 commented:
Regardless of who starts the show, where they are in the world, or if a player or the host leaves or joins the server, all the fireworks will be synced among all clients now.

Sqrl34 committed:
fix: spawns firework prop

Sqrl34 commented:
The firework cake object now spawns correctly. All should be good _hopefully_
